File: cli/python/src/memgo_cli/config/resolve.py
"""配置键解析与不可变更新。"""

from __future__ import annotations

from dataclasses import replace

from memgo_cli.config.models import (
    MemGoConfig,
)

SHORT_KEY_ALIASES: dict[str, str] = {
    "api_key": "platform.api_key",
    "base_url": "platform.base_url",
    "user_email": "platform.user_email",
    "user_id": "defaults.user_id",
    "agent_id": "defaults.agent_id",
    "app_id": "defaults.app_id",
    "run_id": "defaults.run_id",
}


def redact_key(key: str) -> str:
    """显示脱敏后的 API 密钥。"""
    if not key:
        return "(not set)"
    if len(key) <= 8:
        return key[:2] + "***"
    return key[:4] + "..." + key[-4:]


def get_nested_value(config: MemGoConfig, dotted_key: str) -> object:
    """读取点分键名对应字段，未知键返回空值。"""
    path = SHORT_KEY_ALIASES.get(dotted_key, dotted_key).split(".")
    value: object = config
    for part in path:
        if part.startswith("_") or not hasattr(value, "__dataclass_fields__"):
            return None
        if part not in value.__dataclass_fields__:
            return None
        value = getattr(value, part)
    return value


def set_nested_value(config: MemGoConfig, dotted_key: str, value: str) -> MemGoConfig:
    """返回更新指定字段后的新配置；未知键和无效整数明确失败。"""
    path = SHORT_KEY_ALIASES.get(dotted_key, dotted_key).split(".")
    current = get_nested_value(config, dotted_key)
    if current is None or type(current) not in (str, bool, int):
        raise ValueError(f"Unknown config key: {dotted_key}")
    converted: str | bool | int = value
    if isinstance(current, bool):
        converted = value.lower() in ("true", "1", "yes")
    elif isinstance(current, int):
        converted = int(value)
    if len(path) == 1:
        if path[0] != "version" or type(converted) is not int:
            raise ValueError(f"Invalid config key: {dotted_key}")
        return replace(config, version=converted)
    if len(path) != 2:
        raise ValueError(f"Unknown config key: {dotted_key}")
    section = getattr(config, path[0])
    return replace(config, **{path[0]: replace(section, **{path[1]: converted})})

File: cli/node/src/config/resolve.ts
import type { JsonObject } from '../backend/json.js';
import type { MemGoConfig } from './types.js';
/**
 * 隐藏密钥中间内容，仅保留识别所需片段。
 * @param key - 访问服务的 API 密钥。
 * @returns 脱敏后的密钥展示文本。
 */
export function redactKey(key: string): string {
  if (!key) return '(not set)';
  if (key.length <= 8) return `${key.slice(0, 2)}***`;
  return `${key.slice(0, 4)}...${key.slice(-4)}`;
}

/** 配置路径与内存字段的对应表。 */
const KEY_MAP: Record<string, [keyof MemGoConfig, string]> = {
  'platform.api_key': ['platform', 'apiKey'],
  'platform.base_url': ['platform', 'baseUrl'],
  'platform.user_email': ['platform', 'userEmail'],
  'defaults.user_id': ['defaults', 'userId'],
  'defaults.agent_id': ['defaults', 'agentId'],
  'defaults.app_id': ['defaults', 'appId'],
  'defaults.run_id': ['defaults', 'runId'],
  // 短名称别名。
  api_key: ['platform', 'apiKey'],
  base_url: ['platform', 'baseUrl'],
  user_email: ['platform', 'userEmail'],
  user_id: ['defaults', 'userId'],
  agent_id: ['defaults', 'agentId'],
  app_id: ['defaults', 'appId'],
  run_id: ['defaults', 'runId'],
};

/**
 * 按外部点分键名读取配置字段。
 * @param config - 当前配置。
 * @param dottedKey - 持久化配置使用的点分字段名。
 * @returns 对应字段值；未知键返回未定义。
 */
export function getNestedValue(config: MemGoConfig, dottedKey: string): unknown {
  const mapping = KEY_MAP[dottedKey];
  if (!mapping) return undefined;
  const [section, field] = mapping;
  return (config[section] as unknown as JsonObject)[field];
}

/**
 * 返回新配置；未知字段明确报错，不修改输入。
 * @param config - 当前配置。
 * @param dottedKey - 持久化配置使用的点分字段名。
 * @param value - 待处理的字段值。
 * @returns 仅更新指定字段的新配置。
 */
export function setNestedValue(config: MemGoConfig, dottedKey: string, value: string): MemGoConfig {
  const mapping = KEY_MAP[dottedKey];
  if (!mapping) throw new Error(`Unknown config key: ${dottedKey}`);
  const [section, field] = mapping;
  return {
    ...config,
    [section]: { ...(config[section] as object), [field]: value },
  };
}
